### PR TITLE
[TF-TRT] Removing TRT 5 & 6 from the code base - Large Cleanup

### PR DESCRIPTION
**Pull request by DEKHTIARJonathan**
This PR removes TRT 5 & 6 from the codebase and performs a large cleanup / refactoring

### CODE REVIEW

**Comment by DEKHTIARJonathan** [comment]
Unittest results with TRT 8

```bash
INFO: 603 processes: 81 internal, 522 local.
INFO: Build completed successfully, 603 total actions
//tensorflow/compiler/tf2tensorrt:convert_graph_test                     PASSED in 9.1s
//tensorflow/compiler/tf2tensorrt:convert_graph_test_gpu                 PASSED in 8.7s
//tensorflow/compiler/tf2tensorrt:convert_nodes_test                     PASSED in 144.2s
//tensorflow/compiler/tf2tensorrt:convert_nodes_test_gpu                 PASSED in 144.1s
//tensorflow/compiler/tf2tensorrt:segment_test                           PASSED in 0.2s
//tensorflow/compiler/tf2tensorrt:segment_test_gpu                       PASSED in 0.1s
//tensorflow/compiler/tf2tensorrt:tensorrt_test_cc                       PASSED in 4.1s
//tensorflow/compiler/tf2tensorrt:tensorrt_test_cc_gpu                   PASSED in 4.2s
//tensorflow/compiler/tf2tensorrt:trt_allocator_test                     PASSED in 0.1s
//tensorflow/compiler/tf2tensorrt:trt_engine_op_test                     PASSED in 2.9s
//tensorflow/compiler/tf2tensorrt:trt_engine_op_test_gpu                 PASSED in 2.4s
//tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test           PASSED in 2.4s
//tensorflow/compiler/tf2tensorrt:trt_engine_resource_ops_test_gpu       PASSED in 2.2s
//tensorflow/compiler/tf2tensorrt:trt_lru_cache_test                     PASSED in 0.1s
//tensorflow/compiler/tf2tensorrt:trt_shape_optimization_profiles_test   PASSED in 7.0s
//tensorflow/compiler/tf2tensorrt:trt_shape_optimization_profiles_test_gpu PASSED in 7.0s

Executed 16 out of 16 tests: 16 tests pass.

# =============================================================================== #

INFO: 3 processes: 1 internal, 2 local.
INFO: Build completed successfully, 3 total actions
//tensorflow/python/compiler/tensorrt:trt_convert_test                   PASSED in 56.8s
//tensorflow/python/compiler/tensorrt:trt_convert_test_gpu               PASSED in 56.8s

INFO: Build completed successfully, 3 total actions

# =============================================================================== #

Running ./tensorflow/python/compiler/tensorrt/test/annotate_max_batch_sizes_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/base_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/batch_matmul_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/biasadd_matmul_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/binary_tensor_weight_broadcast_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/cast_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/combined_nms_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/concatenation_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/const_broadcast_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/conv2d_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/dynamic_input_shapes_test.py ... [IGNORED]
Running ./tensorflow/python/compiler/tensorrt/test/identity_output_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/int32_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/lru_cache_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/memory_alignment_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/multi_connection_neighbor_engine_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/neighboring_engine_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/quantization_mnist_test.py ... [IGNORED]
Running ./tensorflow/python/compiler/tensorrt/test/quantization_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/rank_two_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/reshape_transpose_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/tf_function_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/topk_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/trt_mode_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/unary_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/vgg_block_nchw_test.py ... [SUCCESS]
Running ./tensorflow/python/compiler/tensorrt/test/vgg_block_test.py ... [SUCCESS]
```

File: tensorflow/python/compiler/tensorrt/__init__.py [hunk 1]
```diff
@@ -18,6 +18,69 @@
...
+        "of TensorRT during compilation and runtime.")
+    raise RuntimeError("Incompatible TensorRT major version")
+
+  if loaded_version != linked_version:
+    logging.info(
+        "Loaded TensorRT %s and linked TensorFlow against TensorRT %s. " %
+        (".".join(str(x) for x in loaded_version), ".".join(
+            str(x) for x in linked_version)) +
+        "This is supported because TensorRT " +
+        " minor/patch upgrades are backward compatible")
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/__init__.py:86]: Similar to the comment before, what is "backward"?

**Review by bixia1 — COMMENTED** [review]
Can you rebase, squash, and also fix the PR description so that it only contains a description of what the PR does and nothing else? The PR description will go to the commit message.

File: third_party/gpus/find_cuda_config.py [hunk 2]
```diff
@@ -520,15 +520,10 @@ def get_header_version(path):
...
-  try:
-    header_path, header_version = _find_header(base_paths, "NvInfer.h",
-                                               required_version,
-                                               get_header_version)
-  except ConfigError:
-    # TensorRT 6 moved the version information to NvInferVersion.h.
-    header_path, header_version = _find_header(base_paths, "NvInferVersion.h",
-                                               required_version,
-                                               get_header_version)
+  # TensorRT 6 moved the version information to NvInferVersion.h.
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply third_party/gpus/find_cuda_config.py:523]: Fair enough, let's remove this comment I had a doubt on how useful it was anyway.
Removed

File: tensorflow/python/compiler/tensorrt/__init__.py [hunk 3]
```diff
@@ -18,6 +18,69 @@
...
+
+  logging.info("Linked TensorRT version: %s" % str(linked_version))
+  logging.info("Loaded TensorRT version: %s" % str(loaded_version))
+
+  def raise_trt_deprecation_warning(version_type, trt_version):
+    if version_type not in ["linked", "loaded"]:
+      raise ValueError(
+          "Incorrect value received for version_type: {}. "
+          "Accepted: ['linked', 'loaded']"
+      )
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/__init__.py:43]: Fixed

File: tensorflow/python/compiler/tensorrt/__init__.py [hunk 4]
```diff
@@ -18,6 +18,69 @@
...
+
+    logging.error(
+        "The {version_type} version of TensorRT: `{trt_version}` has now "
+        "been removed. Please upgrade to TensorRT 7 or more recent.".format(
+            version_type=version_type,
+            trt_version=".".join([str(ver) for ver in trt_version])
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/__init__.py:49]: Fixed

File: tensorflow/python/compiler/tensorrt/__init__.py [hunk 5]
```diff
@@ -18,6 +18,69 @@
...
+
+  if not trt_utils.IsLoadedTensorRTVersionGreaterEqual(7, 0, 0):
+    raise_trt_deprecation_warning("loaded", loaded_version)
+
+  if not trt_utils.IsLoadedTensorRTVersionGreaterEqual(*linked_version):
+    logging.error(
+        "Loaded TensorRT %s but linked TensorFlow against TensorRT %s. " %
+        (".".join(str(x) for x in loaded_version), ".".join(
+            str(x) for x in linked_version)) +
+        "TensorRT does not support forward compatibility. " +
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/__init__.py:66]: I did not invent this phrasing, see here:
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/compiler/tensorrt/trt_convert.py#L228
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/compiler/tensorrt/trt_convert.py#L246

If you want to update it, please rephrase the comment, I'm not exactly sure how to change that

File: tensorflow/python/compiler/tensorrt/__init__.py [hunk 6]
```diff
@@ -18,6 +18,69 @@
...
+    raise RuntimeError("Incompatible TensorRT versions")
+
+  if loaded_version[0] > linked_version[0]:
+    logging.error(
+        "Loaded TensorRT %s but linked TensorFlow against TensorRT %s. " %
+        (".".join(str(x) for x in loaded_version), ".".join(
+            str(x) for x in linked_version)) +
+        "It is required to use the same major version " +
+        "of TensorRT during compilation and runtime.")
+    raise RuntimeError("Incompatible TensorRT major version")
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/__init__.py:78]: Fixed

File: tensorflow/python/compiler/tensorrt/__init__.py (hunk 1 above)
Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/__init__.py:86]: Here: https://github.com/tensorflow/tensorflow/pull/50782#discussion_r672716824

File: tensorflow/python/compiler/tensorrt/trt_convert.py [hunk 7]
```diff
@@ -206,46 +202,6 @@ def _check_conversion_params(conversion_params, is_v2=False):
                      "(to disable main graph conversion).")
 
 
-def _check_trt_version_compatibility():
```
Comment (bixia1) [inline tensorflow/python/compiler/tensorrt/trt_convert.py:209]: Let's split this change to a follow up PR. I meant, the change to move this check to the initialization of the package.

Comment (DEKHTIARJonathan) [reply tensorflow/python/compiler/tensorrt/trt_convert.py:209]: Fixed

**Comment by DEKHTIARJonathan** [comment]
@bixia1 change requested applied - PR squashed

**Review by bixia1 — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Pull request merged**